test/hotspot/jtreg/compiler/vectorapi/TestVectorAddMulReduction.java

```diff
@@ -0,0 +1,184 @@
+        failOn = {"requires_strict_order"}, phase = CompilePhase.PRINT_IDEAL)
+    public static void testDoubleAdd_128() {
+        testDoubleAddKernel(DoubleVector.SPECIES_128, da);
+    }
+
+    @Test
+    @IR(applyIf = {"MaxVectorSize", ">=32"}, counts = {IRNode.ADD_REDUCTION_VD, ">=1", "no_strict_order", ">=1"},
+        failOn = {"requires_strict_order"}, phase = CompilePhase.PRINT_IDEAL)
+    public static void testDoubleAdd_256() {
+        testDoubleAddKernel(DoubleVector.SPECIES_256, da);
+    }
+
+    @Test
+    @IR(applyIf = {"MaxVectorSize", ">=64"}, counts = {IRNode.ADD_REDUCTION_VD, ">=1", "no_strict_order", ">=1"},
+        failOn = {"requires_strict_order"}, phase = CompilePhase.PRINT_IDEAL)
+    public static void testDoubleAdd_512() {
+        testDoubleAddKernel(DoubleVector.SPECIES_512, da);
+    }
+
+    // Test mul reduction operation for floats
+    // On aarch64, there are no direct vector mul reduction instructions for float/double mul reduction
+    // and scalar instructions are emitted for 64-bit/128-bit vectors. Thus MulReductionVF/VD nodes are generated
+    // only for vector length of 8B/16B on vectorAPI.
+    @ForceInline
+    public static void testFloatMulKernel(VectorSpecies SPECIES, float[] f) {
+        for (int i = 0; i < SPECIES.loopBound(f.length); i += SPECIES.length()) {
+            var av = FloatVector.fromArray(SPECIES, f, i);
+            fres += av.reduceLanes(VectorOperators.MUL);
+        }
+    }
+
+    @Test
+    @IR(applyIf = {"MaxVectorSize", ">=8"}, counts = {IRNode.MUL_REDUCTION_VF, ">=1", "no_strict_order", ">=1"},
+        failOn = {"requires_strict_order"}, phase = CompilePhase.PRINT_IDEAL)
+    public static void testFloatMul_64() {
+        testFloatMulKernel(FloatVector.SPECIES_64, fa);
+    }
+
+    @Test
+    @IR(applyIf = {"MaxVectorSize", ">=16"}, counts = {IRNode.MUL_REDUCTION_VF, ">=1", "no_strict_order", ">=1"},
+        failOn = {"requires_strict_order"}, phase = CompilePhase.PRINT_IDEAL)
+    public static void testFloatMul_128() {
+        testFloatMulKernel(FloatVector.SPECIES_128, fa);
+    }
+
+    // Test mul reduction operation for doubles
+    @ForceInline
+    public static void testDoubleMulKernel(VectorSpecies SPECIES, double[] d) {
+        for (int i = 0; i < SPECIES.loopBound(d.length); i += SPECIES.length()) {
+            var av = DoubleVector.fromArray(SPECIES, d, i);
+            dres += av.reduceLanes(VectorOperators.MUL);
+        }
+    }
+
+    @Test
+    @IR(applyIf = {"MaxVectorSize", ">=16"}, counts = {IRNode.MUL_REDUCTION_VD, ">=1", "no_strict_order", ">=1"},
+        failOn = {"requires_strict_order"}, phase = CompilePhase.PRINT_IDEAL)
+    public static void testDoubleMul_128() {
+        testDoubleMulKernel(DoubleVector.SPECIES_128, da);
+    }
+
+    public static void main(String[] args) {
+        TestFramework.runWithFlags("-XX:-TieredCompilation",
```

Comment:
Why `-XX:-TieredCompilation`?

Reply:
I added that to enable c2 compilation but I just realized that this flag is whitelisted already and maybe not needed here.